drivers/interrupt_controller/intc_gicv3.c

```diff
@@ -513,8 +513,11 @@ static mem_addr_t arm_gic_iterate_rdists(void)
 	for (mem_addr_t rdist_addr = GIC_RDIST_BASE;
 		rdist_addr < GIC_RDIST_BASE + GIC_RDIST_SIZE;
 		rdist_addr += 0x20000) {
-		uint64_t val = sys_read64(rdist_addr + GICR_TYPER);
-		uint64_t gicr_aff = GICR_TYPER_AFFINITY_VALUE_GET(val);
+		uint64_t val = sys_read32(rdist_addr + GICR_TYPER + 4);
+		uint64_t gicr_aff = val;
+
+		val <<= GICR_TYPER_AFFINITY_VALUE_SHIFT;
+		val |= sys_read32(rdist_addr + GICR_TYPER);
```

Comment:
Your proposed change reads a bit odd to me. You are meant to read the low and high 32-bit and then composed a 64-bit value. So I would expect the shift to be `32` rather than `GICR_TYPE_AFFINITY_VALUE_SHIFT`.

I understand the define will result to ``32`` as well. But that's less obvious. So I would suggest the following:

```
uint64_t val = sys_read32(rdist_addr + GICR_TYPER);
uint64_t gicr_aff;

val <<= 32;
val |= sys_read32(rdist_addr + GICR_TYPER + 4);

gicr_aff = GICR_TYPER_AFFINITY_VALUE_GET(val);
``` 

That said, I expect all the 64-bit access to GICv3 distributor/re-distributor to require two 32-bits accesses. Furthermore, IMHO, it would be better if the access stay 64-bit for Arm64. So shouldn't we provide a arm32 version of ``sys_read64()`` ?

Reply:
```
uint64_t val = sys_read32(rdist_addr + GICR_TYPER);
uint64_t gicr_aff;

val <<= 32;
val |= sys_read32(rdist_addr + GICR_TYPER + 4);

gicr_aff = GICR_TYPER_AFFINITY_VALUE_GET(val);
```
This seems incorrect to me as you are moving the value read from GICR_TYPER to the upper 32 bits. And the value read from (GICR_TYPER + 4) to the lower 32 bits.

> Furthermore, IMHO, it would be better if the access stay 64-bit for Arm64. So shouldn't we provide a arm32 version of sys_read64() ?
I was hoping that we could restrict splitting sys_read64() into 2 sys_read32() only for emulated MMIO regions (eg GICR, GICD). For other regions, it can stay as sys_read64(), isn't it ?

Reply:
Agreed with @jgrall idea, your proposed change seems to make `gicr_aff` only from the lower 32 bits, it would be wrong for AArch64.

Reply:
> This seems incorrect to me as you are moving the value read from GICR_TYPER to the upper 32 bits. And the value read from (GICR_TYPER + 4) to the lower 32 bits. It should be the other way around.

You are right.

> I was hoping that we could restrict splitting sys_read64() into 2 sys_read32() only for emulated MMIO regions (eg GICR, GICD). For other regions, it can stay as sys_read64(), isn't it ?

In theory any MMIO regions would have the same issue because the HW may not support 64-bit access on MMIO (see [1]). So I think they all should use two 32-bit accesses. 

[1] https://lore.kernel.org/xen-devel/6D2FD203-04D9-4CDD-B57D-E90F5BC5CC3A@arm.com/T/#mcc2b226e506c669918eee0266067f6d746ccc744

Reply:
@jgrall @povergoing 
Please have a look at my latest patch.
I have modified sys_read64() into two sys_read32() for the aarch32 implementation.

Reply:
LGTM, but could it be more compact？ like
```
uint64_t val;

val = sys_read32(addr);
val |= (uint64_t)sys_read32(addr + 4) << 32;
```